src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2134,23 +2144,80 @@ void EmitAnchors(RegexNode node)
             void EmitMultiChar(RegexNode node)
             {
-                // if (textSpanPos + node.Str.Length >= textSpan.Length) goto doneLabel;
-                // if (node.Str[0] != textSpan[textSpanPos]) goto doneLabel;
-                // if (node.Str[1] != textSpan[textSpanPos+1]) goto doneLabel;
-                // ...
-                EmitSpanLengthCheck(node.Str!.Length);
-                for (int i = 0; i < node.Str!.Length; i++)
+                ReadOnlySpan<char> s = node.Str;
+
+                // Emit the length check for the whole string.  If the generated code gets past this point,
+                // we know the span is at least textSpanPos + s.Length long.
+                EmitSpanLengthCheck(s.Length);
+
+                // If we're doing a case-insensitive comparison, we need to lower case each character,
+                // so we just go character-by-character.  But if we're not, we try to process multiple
+                // characters at a time; this is helpful not only for throughput but also in reducing
+                // the amount of IL and asm that results from this unrolling.
+                if (!IsCaseInsensitive(node))
+                {
+                    void EmitTextSpanOffsetAsByte()
+                    {
+                        // Unsafe.As<char, byte>(ref Unsafe.Add(ref MemoryMarshal.GetReference(textSpan), textSpanPos))
+                        Ldloc(textSpanLocal);
+                        Call(s_memoryMarshalGetReference);
+                        if (textSpanPos > 0)
+                        {
+                            Ldc(textSpanPos);
+                            Call(s_unsafeAddChar);
+                        }
+                        Call(s_unsafeCharAsByte);
+                    }
+
+                    // TODO https://github.com/dotnet/corefx/issues/39227:
+                    // If/when we implement CompileToAssembly, this code will either need to be special-cased
+                    // to not be used when saving out the assembly, or it'll need to be augmented to emit an
+                    // endianness check into the IL.  The code below is creating int/long constants based on
+                    // reading the comparison string at compile time, and the machine doing the compilation
+                    // could be of a different endianness than the machine running the compiled assembly.
+
+                    // On 64-bit, process 4 characters at a time until the string isn't at least 4 characters long.
+                    if (IntPtr.Size == 8)
+                    {
+                        const int CharsPerInt64 = 4;
+                        while (s.Length >= CharsPerInt64)
+                        {
+                            // if (Unsafe.ReadUnaligned<long>(ref text) != value) goto doneLabel;
+                            EmitTextSpanOffsetAsByte();
+                            Call(s_unsafeReadUnalignedInt64);
```

Comment:
You can just emit the the regular ldind.i8 instruction with unaligned prefix here. No need to go through the troubles of referencing `s_unsafeReadUnalignedInt64` and getting it inlined. (Same for the other ones.)

Reply:
Great.